lib/src/main/java/com/scalar/admin/kubernetes/Pauser.java

```diff
@@ -77,67 +103,188 @@ public PausedDuration pause(int pauseDuration, @Nullable Long maxPauseWaitTime)
     }
 
-    return new PausedDuration(startTime, endTime);
+    // We use the exceptions as conditions instead of using boolean flags like `isPauseOk`, etc. If
+    // we use boolean flags, it might cause a bit large number of combinations. For example, if we
+    // have three flags, they generate 2^3 = 8 combinations. It also might make the nested if
+    // statement or a lot of branches of the switch statement. That's why we don't use status flags
+    // for now.
+    PauserException pauserException =
+        buildException(
+            unpauseFailedException,
+            pauseFailedException,
+            getTargetAfterPauseFailedException,
+            statusCheckFailedException,
+            statusUnmatchedException);
+
+    // Return the final result based on each process.
+    if (pauserException != null) {
+      // Some operations failed.
+      throw pauserException;
+    } else {
+      // All operations succeeded.
+      return pausedDuration;
+    }
   }
 
-  private void unpauseWithRetry(
-      RequestCoordinator coordinator, int maxRetryCount, TargetSnapshot target) {
+  @VisibleForTesting
+  void unpauseWithRetry(RequestCoordinator coordinator, int maxRetryCount) {
     int retryCounter = 0;
-
     while (true) {
       try {
         coordinator.unpause();
         return;
       } catch (Exception e) {
         if (++retryCounter >= maxRetryCount) {
-          logger.warn(
-              "Failed to unpause Scalar product. They are still in paused. You must restart related"
-                  + " pods by using the `kubectl rollout restart deployment {}`"
-                  + " command to unpase all pods.",
-              target.getDeployment().getMetadata().getName());
-          return;
+          throw e;
         }
       }
     }
   }
 
+  @VisibleForTesting
+  TargetSnapshot getTarget() throws PauserException {
+    return targetSelector.select();
+  }
+
+  @VisibleForTesting
   RequestCoordinator getRequestCoordinator(TargetSnapshot target) {
     return new RequestCoordinator(
         target.getPods().stream()
             .map(p -> new InetSocketAddress(p.getStatus().getPodIP(), target.getAdminPort()))
             .collect(Collectors.toList()));
   }
+
+  @VisibleForTesting
+  PausedDuration pauseInternal(
+      RequestCoordinator requestCoordinator, int pauseDuration, @Nullable Long maxPauseWaitTime) {
+    requestCoordinator.pause(true, maxPauseWaitTime);
+    Instant startTime = Instant.now();
+    Uninterruptibles.sleepUninterruptibly(pauseDuration, TimeUnit.MILLISECONDS);
+    Instant endTime = Instant.now();
+    return new PausedDuration(startTime, endTime);
+  }
+
+  @VisibleForTesting
+  @Nullable
+  StatusUnmatchedException targetStatusEquals(
+      TargetSnapshot before, @Nullable TargetSnapshot after) {
+    if (after != null && before.getStatus().equals(after.getStatus())) {
+      return null;
+    } else {
+      return new StatusUnmatchedException(STATUS_UNMATCHED_ERROR_MESSAGE);
```

Comment:
[minor] How about adding `before` and `after` to the error message just in case?

Reply:
Ah, I see. It might have a lot of information that makes the error message very long, but I will check what is included in `before` and `after`, first. And, if the information is not too large, I will add them into the error message.

Thank you for your suggestion!

Reply:
I checked what is included in `TargetStatus` (i.e., `before` and `after`). As a result, I think we can add their information to the log, but in the current implementation, I think the information might not be readable a little.

So, I think it would be better to implement `TargetStatus.toString()` first. However, I think we need to consider something before implementing it.

Therefore, I will keep the current implementation as is in this PR, and I will fix it in another PR in the future.